lib/Pinto/Shell.pm

```diff
@@ -0,0 +1,63 @@
+# ABSTRACT: Shell into a distribution
+
+package Pinto::Shell;
+
+use Moose;
+use MooseX::StrictConstructor;
+use MooseX::MarkAsMethods ( autoclean => 1 );
+use MooseX::Types::Moose qw(Str);
+
+use Pinto::Types qw(File Dir);
+use Pinto::Util qw(debug throw);
+use Cwd::Guard qw(cwd_guard);
+
+#------------------------------------------------------------------------------
+
+# VERSION
+
+#------------------------------------------------------------------------------
+
+has shell => (
+    is       => 'ro',
+    isa      => File,
+    required => 1,
+);
+
+has archive => (
+    is       => 'ro',
+    isa      => File,
+    required => 1,
+);
+
+has unpacker => (
+    is       => 'ro',
+    isa      => 'Pinto::ArchiveUnpacker',
+    default  => sub { Pinto::ArchiveUnpacker->new( archive => $_[0]->archive ) },
+    init_arg => undef,
+    lazy     => 1,
+);
+
+has work_dir => (
+    is       => 'ro',
+    isa      => Dir,
+    default  => sub { $_[0]->unpacker->unpack },
+    init_arg => undef,
+    lazy     => 1,
+);
+
+#------------------------------------------------------------------------------
+
+sub spawn {
+    my ($self) = @_;
+
+    my $shell = $self->shell;
+
+    my $cwd_guard = cwd_guard( $self->work_dir );
```

Comment:
Lovely :-)


Kahlil (Kal) Hodgson                       GPG: C9A02289
Head of Technology                         (m) +61 (0) 4 2573 0382
DealMax Pty Ltd

Suite 1416
401 Docklands Drive
Docklands VIC 3008 Australia

"All parts should go together without forcing.  You must remember that
the parts you are reassembling were disassembled by you.  Therefore,
if you can't get them together again, there must be a reason.  By all
means, do not use a hammer."  -- IBM maintenance manual, 1925

On 6 February 2015 at 16:32, Jeffrey Ryan Thalhammer <
notifications@github.com> wrote:

> In lib/Pinto/Shell.pm
> <https://github.com/thaljef/Pinto/pull/184#discussion_r24223481>:
>
> > +has work_dir => (
> > +    is       => 'ro',
> > +    isa      => Dir,
> > +    default  => sub { $_[0]->unpacker->unpack },
> > +    init_arg => undef,
> > +    lazy     => 1,
> > +);
> > +
> > +#------------------------------------------------------------------------------
> > +
> > +sub spawn {
> > +    my ($self) = @_;
> > +
> > +    my $shell = $self->shell;
> > +
> > +    my $cwd_guard = cwd_guard( $self->work_dir );
>
> That worked out quite nicely, don't you think?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/thaljef/Pinto/pull/184/files#r24223481>.
>